### PR TITLE
Sema: Optimize ConstraintSystem::is{Decl,Conformance}Unavailable

### PR DESCRIPTION
**Pull request by slavapestov**
Let's avoid creating an ExportContext, which computes a bunch of
irrelevant stuff. Also, delay the expensive call to
overApproximateAvailabilityAtLocation() unless we know the
declaration is conditionally unavailable.

### CODE REVIEW

File: lib/Sema/ConstraintSystem.cpp [hunk 1]
```diff
@@ -5089,17 +5089,39 @@ bool ConstraintSystem::isDeclUnavailable(const Decl *D,
...
-  return result.hasValue();
+  AvailabilityContext runningOSOverApprox =
+      TypeChecker::overApproximateAvailabilityAtLocation(loc, DC);
+
+  // The reference is safe if an over-approximation of the running OS
+  // versions is fully contained within an under-approximation
+  // of the versions on which the declaration is available. If this
+  // containment cannot be guaranteed, we say the reference is
+  // not available.
+  if (runningOSOverApprox.isContainedIn(safeRangeUnderApprox))
```
Comment (xedin) [inline lib/Sema/ConstraintSystem.cpp:5121]: Nit: `return !runningOSOverApprox.isContainedIn(safeRangeUnderApprox);`

**Review by xedin — COMMENTED** [review]
This is almost exactly `checkDeclarationAvailability` besides availability reason computation at the end and `isAwaysAvailable` bit (does it make sense to add it there?). Should we just add a new method `checkDeclarationAvailability` which doesn't require `ExprContext` and returns a bool?

**Comment by slavapestov** [comment]
@xedin It takes a ConstraintLocator too (notice how obtaining the source location is delayed as much as possible). Nothing else in TypeCheckAvailability.cpp deals with locators or other constraint system stuff. Are you sure it makes sense to put it there instead?

File: lib/Sema/ConstraintSystem.cpp (hunk 1 above)
Comment (slavapestov) [reply lib/Sema/ConstraintSystem.cpp:5121]: Ah, heh. Fixed

**Comment by slavapestov** [comment]
@swift-ci Please smoke test

**Comment by slavapestov** [comment]
@swift-ci Please test source compatibility

**Comment by xedin** [comment]
@slavapestov Getting SourceLoc from ASTNode is not an expensive operation, new method can just accept ASTNode as an argument. The main concern for me is that we'd have almost exactly the same logic in two different places.

**Pull request merged**